test/tests_for_sdkmr/job_submission_parameters_test.py

```diff
@@ -207,24 +228,256 @@ def test_hash():
-    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1)) == hash(
-        JobRequirements(1, 1, 1, c1a, t, u1a, f, r1a)
+    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1, t)) == hash(
+        JobRequirements(1, 1, 1, c1a, t, u1a, f, r1a, t)
     )
 
     assert hash(JobRequirements(1, 1, 1, c1)) != hash(JobRequirements(2, 1, 1, c1a))
     assert hash(JobRequirements(1, 1, 1, c1)) != hash(JobRequirements(1, 2, 1, c1a))
     assert hash(JobRequirements(1, 1, 1, c1)) != hash(JobRequirements(1, 1, 2, c1a))
     assert hash(JobRequirements(1, 1, 1, c1)) != hash(JobRequirements(1, 1, 1, c2))
 
-    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1)) != hash(
-        JobRequirements(1, 1, 1, c1a, f, u1a, f, r1a)
+    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1, t)) != hash(
+        JobRequirements(1, 1, 1, c1a, f, u1a, f, r1a, t)
+    )
+    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1, t)) != hash(
+        JobRequirements(1, 1, 1, c1a, t, u2, f, r1a, t)
+    )
+    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1, t)) != hash(
+        JobRequirements(1, 1, 1, c1a, t, u1a, t, r1a, t)
+    )
+    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1, t)) != hash(
+        JobRequirements(1, 1, 1, c1a, t, u1a, f, r2, t)
+    )
+    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1, t)) != hash(
+        JobRequirements(1, 1, 1, c1a, t, u1a, f, r1a, f)
+    )
+
+
+def test_job_sub_init_minimal():
+    jsp = JobSubmissionParameters(
+        "jobid",
+        AppInfo("a.b", "a/x"),
+        JobRequirements(6, 7, 4, "cligroup"),
+        UserCreds("user", "tokeytoken"),
+    )
+
+    assert jsp.job_id == "jobid"
+    assert jsp.app_info == AppInfo("a.b", "a/x")
+    assert jsp.job_reqs == JobRequirements(6, 7, 4, "cligroup")
+    assert jsp.user_creds == UserCreds("user", "tokeytoken")
+    assert jsp.parent_job_id is None
+    assert jsp.wsid is None
+    assert jsp.source_ws_objects == tuple()
+
+
+def test_job_sub_init_maximal():
+    jsp = JobSubmissionParameters(
+        "    jobid  \t  ",
+        AppInfo("a.b", "a/x"),
+        JobRequirements(6, 7, 4, "cligroup"),
+        UserCreds("user", "tokeytoken"),
+        "    parentid  \t   ",
+        1,
+        ["   1   /\t2   /   4", "6/7/8"],
     )
-    assert hash(JobRequirements(1, 1, 1, c1, t, u1, f, r1)) != hash(
-        JobRequirements(1, 1, 1, c1a, t, u2, f, r1a)
+
+    assert jsp.job_id == "jobid"
+    assert jsp.app_info == AppInfo("a.b", "a/x")
+    assert jsp.job_reqs == JobRequirements(6, 7, 4, "cligroup")
+    assert jsp.user_creds == UserCreds("user", "tokeytoken")
+    assert jsp.parent_job_id == "parentid"
+    assert jsp.wsid == 1
+    assert jsp.source_ws_objects == ("1/2/4", "6/7/8")
+
+
+def test_job_sub_init_fail():
+    n = None
+    j = "jobby job job"
+    a = AppInfo("a.b", "a/x")
+    r = JobRequirements(6, 7, 4, "cligroup")
+    u = UserCreds("user", "tokeytoken")
```

Comment:
no `tokey McTokenFace`?

Reply:
I'm not a fan of the XyMcXyFace meme. One's sense of humor must rise above, must it not?

Reply:
It must indeed!